woocommerce/woocommerce-ios
# [Migrate simple payments] Show bottom sheet that adds custom amount

Parts of #12387 
⚠️ Please make sure https://github.com/woocommerce/woocommerce-ios/pull/12388 is approved first if it hasn't been merged


## Why

As we're migrating from simple payments to order creation, we want to show the order form with a bottom sheet instead of the simple payments when the user taps on Payments > Collect Payment. After https://github.com/woocommerce/woocommerce-ios/pull/12388, this PR implements the migration UI and presents it in a half sheet with the CTA to dismiss the sheet then present the custom amount sheet. Please note that I'm leaving the "collect payment" flow from the order form as a separate subtask, because it turns out to require more changes as this flow doesn't use any hosting controller.

## How

The migration UI is in `SimplePaymentsMigrationView` with a new image `bell` exported from the design.

3 states were added to `InPersonPaymentsMenuViewModel`, whose life cycle is the same as the menu tab to achieve the following behavior:
- The migration sheet is shown **once** whenever the order form is shown when the user taps on `Collect Payment`: this state is tracked in `hasPresentedCollectPaymentMigrationSheet`, while `presentCollectPaymentMigrationSheet` is bound to the migration sheet which can be set to `false` automatically when the sheet is dismissed. `onAppear` is triggered similar to `viewWillAppear`, and we need to make sure the migration sheet is only shown once per payment collection session
- When the user taps on `Add a Custom Amount` in the migration sheet, we want to dismiss the migration sheet then present the custom amount sheet. However, if we just set the migration sheet presentation boolean state to `false` and the custom amount sheet presentation boolean state to `true`, it results in a navigation error like `Attempt to present <_TtGC7SwiftUI29PresentationHostingControllerVS_7AnyView_: 0x121d7e400> on <WooCommerce.MainTabBarController: 0x12005e400> (from <_TtGC7SwiftUI19UIHostingControllerVVS_P10$1c5c8538c21BridgedNavigationView8RootView_: 0x121e03800>) which is already presenting <_TtGC7SwiftUI29PresentationHostingControllerVS_7AnyView_: 0x121895800>.` because the migration sheet hasn't been fully dismissed while trying to present the custom amount sheet. This is resolved by introducing another state `presentCustomAmountAfterDismissingCollectPaymentMigrationSheet` that tracks whether the user taps `Add a Custom Amount` and implementing `onDismiss` of the migration sheet to check for this state

With the 3 boolean states, it's easy to miss resetting some of their values so I added some test cases.

## Testing instructions


- [x] @jaclync tests when the feature flag is off

---

- Go to the menu tab
- Tap on Payments > Collect Payment --> the order creation form should be shown, and a sheet should appear shortly with info about migrating simple payments to order creation. This sheet is a half sheet in compact horizontal size class, and an overlay sheet in regular horizontal size class
- Tap outside to dismiss the modal --> it should focus back on the order form
- Tap on another tab (e.g. Orders tab), and then back to the Menu tab --> the migration sheet shouldn't be shown
- Tap < to go back to the payments
- Tap `Collect Payment` again --> the order creation form should be shown, and the migration sheet should appear shortly
- Tap `Add a custom amount` --> the migration sheet should be dismissed then the custom amount sheet for should be shown
- Enter some custom amount
- Tap `Add Custom Amount`
- Tap `Create` to create the order --> the order form should be popped back to the payments view
- Tap on the Orders tab --> the newly created order with the custom amount should be shown

## Screenshots





https://github.com/woocommerce/woocommerce-ios/assets/1945542/41345e2c-8431-4d5c-adad-0b759a51e7b4

light | dark
-- | --
<img src="https://github.com/woocommerce/woocommerce-ios/assets/1945542/9d16ae63-b0da-4f83-a2a9-6bb96905255a" width="300" /> | <img src="https://github.com/woocommerce/woocommerce-ios/assets/1945542/6534d15d-8757-4124-b7d5-3c53d3d9d365" width="300" />





---
